src/backend/src/migrate.rs

```diff
@@ -67,8 +77,110 @@ pub fn bulk_up(data: &[u8]) {
     }
 }
 
-#[allow(clippy::unused_async)] // TODO: remove this once we make real function calls
-pub async fn step_migration() {
+/// The next chunk of user tokens to be migrated.
+fn next_user_token_chunk(last_user_token: Option<Principal>) -> Vec<(Principal, Vec<UserToken>)> {
+    let chunk_size = 5;
+    let range = last_user_token.map_or((Bound::Unbounded, Bound::Unbounded), |token| {
+        (Bound::Excluded(StoredPrincipal(token)), Bound::Unbounded)
+    });
+    read_state(|state| {
+        state
+            .user_token
+            .range(range)
+            .take(chunk_size)
+            .map(|(stored_principal, token)| (stored_principal.0, token.0))
+            .collect::<Vec<_>>()
+    })
+}
+
+/// The next chunk of custom tokens to be migrated.
+fn next_custom_token_chunk(
+    last_custom_token: Option<Principal>,
+) -> Vec<(Principal, Vec<CustomToken>)> {
+    let chunk_size = 5;
+    let range = last_custom_token.map_or((Bound::Unbounded, Bound::Unbounded), |token| {
+        (Bound::Excluded(StoredPrincipal(token)), Bound::Unbounded)
+    });
+    read_state(|state| {
+        state
+            .custom_token
+            .range(range)
+            .take(chunk_size)
+            .map(|(stored_principal, token)| (stored_principal.0, token.0))
+            .collect::<Vec<_>>()
+    })
+}
+
+/// The next chunk of user profiles to be migrated.
+fn next_user_profile_chunk(
+    last_user_profile: Option<(Timestamp, Principal)>,
+) -> Vec<((Timestamp, Principal), StoredUserProfile)> {
+    let chunk_size = 5;
+    let range = last_user_profile.map_or(
+        (Bound::Unbounded, Bound::Unbounded),
+        |(timestamp, principal)| {
+            (
+                Bound::Excluded((timestamp, StoredPrincipal(principal))),
+                Bound::Unbounded,
+            )
+        },
+    );
+    read_state(|state| {
+        state
+            .user_profile
+            .range(range)
+            .take(chunk_size)
+            .map(|((timestamp, stored_principal), profile)| {
+                ((timestamp, stored_principal.0), profile.0)
+            })
+            .collect::<Vec<_>>()
+    })
+}
+
+/// The next chunk of user timestamps to be migrated.
+fn next_user_timestamp_chunk(user_maybe: Option<Principal>) -> Vec<(Principal, Timestamp)> {
+    let chunk_size = 5;
+    let range = user_maybe.map_or((Bound::Unbounded, Bound::Unbounded), |user| {
+        (Bound::Excluded(StoredPrincipal(user)), Bound::Unbounded)
+    });
+    read_state(|state| {
+        state
+            .user_profile_updated
+            .range(range)
+            .take(chunk_size)
+            .map(|(stored_principal, timestamp)| (stored_principal.0, timestamp))
+            .collect::<Vec<_>>()
+    })
+}
+
+/// Migrates a chunk of data.
+///
+/// # Returns
+/// The updated progress.
+///
+/// # Errors
+/// - Throws a `MigrationError::DataMigrationFailed` if the data transfer fails.
+macro_rules! migrate {
```

Comment:
Interesting. I would have approached that using generics, but this is probably cleaner. 👍

Reply:
I think generics are easier to debug than macros and more familiar to most people.  So I tend to avoid macros where possible.  But I am not sure it is possible to pass say `$chunk_variant` in a generic function, so I _think_ this is the only option.  I didn't spend a long time deciding though; I could be wrong.